Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp

```diff
@@ -165,14 +164,54 @@ void GStreamerVideoEncoder::close()
     m_internalEncoder->close();
 }
 
-GStreamerInternalVideoEncoder::GStreamerInternalVideoEncoder(const String& codecName, VideoEncoder::OutputCallback&& outputCallback, VideoEncoder::PostTaskCallback&& postTaskCallback)
+GStreamerInternalVideoEncoder::GStreamerInternalVideoEncoder(const String& codecName, VideoEncoder::DescriptionCallback&& descriptionCallback, VideoEncoder::OutputCallback&& outputCallback, VideoEncoder::PostTaskCallback&& postTaskCallback)
     : m_codecName(codecName)
+    , m_descriptionCallback(WTFMove(descriptionCallback))
     , m_outputCallback(WTFMove(outputCallback))
     , m_postTaskCallback(WTFMove(postTaskCallback))
 {
     GRefPtr<GstElement> element = gst_element_factory_make("webkitvideoencoder", nullptr);
-    m_harness = GStreamerElementHarness::create(WTFMove(element), [protectedThis = Ref { *this }, this](auto&, const GRefPtr<GstBuffer>& outputBuffer) {
-        if (protectedThis->m_isClosed)
+
+    auto weakThis = WeakPtr { *this };
+    auto pad = adoptGRef(gst_element_get_static_pad(element.get(), "src"));
+    g_signal_connect(pad.get(), "notify::caps", G_CALLBACK(+[](GObject* pad, GParamSpec*, gpointer userData) {
+        auto* encoder = static_cast<GStreamerInternalVideoEncoder*>(userData);
+        if (!encoder)
+            return;
+
+        GRefPtr<GstCaps> caps;
+        g_object_get(pad, "caps", &caps.outPtr(), nullptr);
+        encoder->postTask([encoder, weakEncoder = WeakPtr { *encoder }, caps = WTFMove(caps)] {
+            if (!weakEncoder)
+                return;
+
+            VideoEncoder::ActiveConfiguration configuration;
+            configuration.colorSpace = videoColorSpaceFromCaps(caps.get());
+
+            auto structure = gst_caps_get_structure(caps.get(), 0);
+            GstBuffer* header = nullptr;
+            if (auto streamHeader = gst_structure_get_value(structure, "streamheader")) {
+                RELEASE_ASSERT(GST_VALUE_HOLDS_ARRAY(streamHeader));
+                auto firstValue = gst_value_array_get_value(streamHeader, 0);
+                RELEASE_ASSERT(GST_VALUE_HOLDS_BUFFER(firstValue));
+                header = gst_value_get_buffer(firstValue);
+            } else if (auto codecData = gst_structure_get_value(structure, "codec_data")) {
+                RELEASE_ASSERT(GST_VALUE_HOLDS_BUFFER(codecData));
+                header = gst_value_get_buffer(codecData);
+            }
+
+            if (header) {
+                GstMappedBuffer buffer(header, GST_MAP_READ);
+                configuration.description = { { buffer.data(), buffer.size() } };
+            }
+            encoder->m_descriptionCallback(WTFMove(configuration));
+        });
+    }), weakThis.get());
```

Comment:
Please, correct me if I am wrong but I think this is not wrong, but useless. You're creating here a weak ptr that will always be alive or we would be crashing because of the invalid object we are using.

I guess the intention is creating a weak ptr and passing it to the signal callback. For that you would need to new the WeakPtr and pass the WeakPtr* as userData. Inside the callback you can cast the gpointer to a WeakPtr*, then check for the pointer inside it, get it out to a raw pointer and delete the WeakPtr.

Since you're creating another weak ptr inside the callback to move it into the lambda, I guess instead of newing and destrotying it would be better to do something like, casting to WeakPtr*, move it inside a stack WeakPtr object and then deleting the original WeakPtr*. With the new WeakPtr, you can already extract the contained pointer and move it into the lambda.

Reply:
Ah you mean pass `new WeakPtr { *this }` as userdata?

Reply:
Yes.